subprojects/plugins/src/main/java/org/gradle/api/tasks/GroovyRuntime.java

```diff
@@ -91,15 +97,31 @@ public FileCollection createDelegate() {
                     throw new GradleException(String.format("Cannot infer Groovy class path because no Groovy Jar was found on class path: %s", Iterables.toString(classpath)));
                 }
 
-                if (groovyJar.isGroovyAll() || groovyJar.getVersion().getMajor() == 3) {
+                if (groovyJar.isGroovyAll()) {
                     return project.getLayout().files(groovyJar.getFile());
                 }
 
+                VersionNumber groovyVersion = groovyJar.getVersion();
+                // Groovy 3 does not have groovy-all yet we may have the required pieces on classpath via localGroovy()
+                if (groovyVersion.getMajor() == 3) {
+                    Set<String> groovyLibNames = GROOVY3_LIBS.stream()
+                        .map(libName -> libName + "-" + groovyVersion.toString() + ".jar")
+                        .collect(Collectors.toSet());
+                    List<File> groovyClasspath = new ArrayList<>();
+                    for (File file : classpath) {
+                        if (groovyLibNames.contains(file.getName())) {
+                            groovyClasspath.add(file);
+                        }
+                    }
+                    if (groovyClasspath.size() == GROOVY3_LIBS.size()) {
+                        return project.getLayout().files(groovyClasspath);
+                    }
+                }
```

Comment:
I think this whole block belongs to the same logic (or actually replaces) the log in 133-137. I think we can replace below logic with this

Reply:
Extracted this and the logic below to a separate Groovy3 specific method.